File: src/camoufox_cli/browser.py
"""Browser manager: launches and manages Camoufox instance."""

from __future__ import annotations

import json
import os
import shutil
from datetime import datetime, timezone
from pathlib import Path
from urllib.parse import urlparse

from camoufox.sync_api import Camoufox
from playwright.sync_api import BrowserContext, Page, Request, Response

from .config import CAPSOLVER_XPI_PATH, has_capsolver_xpi, read_config
from .refs import RefRegistry

_MAX_REQUEST_BUFFER = 500


def _ensure_browser_installed() -> None:
    """Check that the Camoufox browser binary is installed, raise if not."""
    try:
        from camoufox.pkgman import get_path
        get_path("camoufox")
    except Exception:
        raise RuntimeError(
            "Browser not found. Run `camoufox-cli install` to download it."
        )


def _install_capsolver_to_profile(profile_path: str) -> None:
    """Copy CapSolver XPI into the Firefox profile's extensions directory."""
    cfg = read_config()
    if not cfg.get("capsolver_extension_id") or not has_capsolver_xpi():
        return

    ext_dir = Path(profile_path) / "extensions"
    ext_dir.mkdir(parents=True, exist_ok=True)
    dest_xpi = ext_dir / f"{cfg['capsolver_extension_id']}.xpi"
    if not dest_xpi.exists():
        shutil.copy2(str(CAPSOLVER_XPI_PATH), str(dest_xpi))


def _configure_capsolver_key(context: BrowserContext, profile_path: str) -> None:
    """
    Configure CapSolver API key in the extension after browser launch.
    Reads the Firefox-assigned UUID from extensions-uuid.json, navigates to
    the extension's options page, and injects the key via chrome.storage / localStorage.
    """
    cfg = read_config()
    if not cfg.get("capsolver_api_key") or not cfg.get("capsolver_extension_id"):
        return

    uuid_map_path = os.path.join(profile_path, "extensions-uuid.json")
    if not os.path.exists(uuid_map_path):
        return  # Extension not yet registered by Firefox

    try:
        with open(uuid_map_path) as f:
            uuid_map = json.load(f)
        uuid = uuid_map.get(cfg["capsolver_extension_id"])
        if not uuid:
            return
    except Exception:
        return

    page = context.new_page()
    try:
        page.goto(
            f"moz-extension://{uuid}/options.html",
            timeout=5000,
            wait_until="domcontentloaded",
        )
        # Inject the API key via chrome.storage and localStorage (both used by CapSolver)
        page.evaluate(
            """(key) => {
                const store =
                    (typeof chrome !== 'undefined' && chrome?.storage?.local) ||
                    (typeof browser !== 'undefined' && browser?.storage?.local);
                if (store) store.set({ apiKey: key });
                localStorage.setItem('capsolver_apikey', key);
                localStorage.setItem('apiKey', key);
            }""",
            cfg["capsolver_api_key"],
        )
    except Exception:
        # Options page not found or unavailable — non-fatal, user can configure manually
        pass
    finally:
        try:
            page.close()
        except Exception:
            pass


class BrowserManager:
    def __init__(self, persistent: str | None = None, proxy: str | None = None):
        self._camoufox: Camoufox | None = None
        self._context: BrowserContext | None = None
        self._page: Page | None = None
        self.refs = RefRegistry()
        self._headless: bool = True
        self._persistent = persistent
        self._proxy = proxy
        # Camoufox spoofs history API for anti-fingerprinting,
        # so we track navigation history ourselves.
        self._history: list[str] = []
        self._history_index: int = -1
        # Network capture
        self._request_buffer: list[dict] = []
        self._pending_requests: dict[int, tuple[Request, dict]] = {}
        self._request_counter: int = 0

    def _on_request(self, request: Request) -> None:
        entry: dict = {
            "id": self._request_counter,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "method": request.method,
            "url": request.url,
            "resourceType": request.resource_type,
            "requestHeaders": dict(request.headers),
            "requestBody": request.post_data,
        }
        self._request_counter += 1
        self._pending_requests[id(request)] = (request, entry)
        self._request_buffer.append(entry)
        if len(self._request_buffer) > _MAX_REQUEST_BUFFER:
            self._request_buffer.pop(0)

    def _on_response(self, response: Response) -> None:
        key = id(response.request)
        pair = self._pending_requests.pop(key, None)
        if pair:
            _, entry = pair
            entry["status"] = response.status
            entry["statusText"] = response.status_text
            entry["responseHeaders"] = dict(response.headers)

    def _setup_network_capture(self, context: BrowserContext) -> None:
        context.on("request", self._on_request)
        context.on("response", self._on_response)

    def get_requests(self, filter_str: str | None = None, n: int | None = None) -> list[dict]:
        result = [r for r in self._request_buffer if filter_str is None or filter_str in r["url"]]
        if n is not None and n > 0:
            result = result[-n:]
        return result

    def clear_requests(self) -> None:
        self._request_buffer.clear()
        self._pending_requests.clear()

    def launch(self, headless: bool = True) -> None:
        if self._camoufox is not None:
            return
        self._headless = headless

        _ensure_browser_installed()

        kwargs: dict = {"headless": headless}
        if self._proxy:
            parsed = urlparse(self._proxy)
            if parsed.scheme not in ("http", "https"):
                raise ValueError(
                    f"Unsupported proxy scheme: {parsed.scheme}. Only http:// and https:// proxies are supported."
                )
            if not parsed.hostname:
                raise ValueError(
                    f"Invalid proxy URL: {self._proxy}. Expected format: http://host:port"
                )
            host_port = parsed.hostname
            if parsed.port:
                host_port += f":{parsed.port}"
            proxy: dict = {"server": f"{parsed.scheme}://{host_port}"}
            if parsed.username:
                proxy["username"] = parsed.username
            if parsed.password:
                proxy["password"] = parsed.password
            kwargs["proxy"] = proxy
        if self._persistent:
            _install_capsolver_to_profile(self._persistent)
            kwargs["persistent_context"] = True
            kwargs["user_data_dir"] = self._persistent

        self._camoufox = Camoufox(**kwargs)
        result = self._camoufox.__enter__()

        if self._persistent:
            # persistent_context returns BrowserContext directly
            self._context = result
            pages = self._context.pages
            self._page = pages[0] if pages else self._context.new_page()

            _configure_capsolver_key(self._context, self._persistent)
        else:
            # Normal mode: result is Browser, new_page() creates default context + page
            self._page = result.new_page()
            self._context = self._page.context

        self._setup_network_capture(self._context)

    def get_page(self) -> Page:
        if self._page is None:
            raise RuntimeError("Browser not launched. Send 'open' command first.")
        return self._page

    def get_context(self) -> BrowserContext:
        if self._context is None:
            raise RuntimeError("Browser not launched. Send 'open' command first.")
        return self._context

    def get_tabs(self) -> list[dict]:
        ctx = self.get_context()
        tabs = []
        for i, p in enumerate(ctx.pages):
            tabs.append({
                "index": i,
                "url": p.url,
                "title": p.title(),
                "active": p is self._page,
            })
        return tabs

    def switch_to_tab(self, index: int) -> Page:
        ctx = self.get_context()
        pages = ctx.pages
        if index < 0 or index >= len(pages):
            raise IndexError(f"Tab index {index} out of range (0-{len(pages) - 1})")
        self._page = pages[index]
        self._page.bring_to_front()
        return self._page

    def close_current_tab(self) -> None:
        ctx = self.get_context()
        pages = ctx.pages
        if len(pages) <= 1:
            raise RuntimeError("Cannot close the last tab. Use 'close' to shut down the browser.")
        current = self._page
        # Switch to another tab before closing
        idx = pages.index(current)
        new_idx = idx - 1 if idx > 0 else 1
        self._page = pages[new_idx]
        self._page.bring_to_front()
        current.close()

    def push_history(self, url: str) -> None:
        """Record a URL in our navigation history."""
        # Truncate forward history when navigating to a new page
        self._history = self._history[:self._history_index + 1]
        self._history.append(url)
        self._history_index = len(self._history) - 1

    def go_back(self) -> str | None:
        """Go back in history. Returns the URL or None if at start."""
        if self._history_index <= 0:
            return None
        self._history_index -= 1
        url = self._history[self._history_index]
        self.get_page().goto(url, wait_until="domcontentloaded")
        return url

    def go_forward(self) -> str | None:
        """Go forward in history. Returns the URL or None if at end."""
        if self._history_index >= len(self._history) - 1:
            return None
        self._history_index += 1
        url = self._history[self._history_index]
        self.get_page().goto(url, wait_until="domcontentloaded")
        return url

    def close(self) -> None:
        if self._camoufox is not None:
            try:
                self._camoufox.__exit__(None, None, None)
            except Exception:
                pass
            self._camoufox = None
            self._context = None
            self._page = None
            self._history.clear()
            self._history_index = -1
            self._request_buffer.clear()
            self._pending_requests.clear()

    @property
    def is_running(self) -> bool:
        return self._camoufox is not None

File: js/src/browser.ts
/** Browser manager: launches and manages Camoufox instance. */

import * as fs from "node:fs";
import * as path from "node:path";
import { execFileSync } from "node:child_process";
import { Camoufox, launchOptions } from "camoufox-js";
import { firefox, type Browser, type BrowserContext, type Page } from "playwright-core";
import { RefRegistry } from "./refs.js";
import { readConfig, CAPSOLVER_XPI_PATH, hasCapsolverXpi } from "./config.js";

function ensureBrowserInstalled(): void {
  try {
    execFileSync("npx", ["camoufox-js", "path"], { stdio: "pipe" });
  } catch {
    throw new Error(
      "Browser not found. Run `camoufox-cli install` to download it."
    );
  }
}

/** Copy CapSolver XPI into the Firefox profile's extensions directory. */
function installCapsolverToProfile(profilePath: string): void {
  const cfg = readConfig();
  if (!cfg.capsolver_extension_id || !hasCapsolverXpi()) return;

  const extDir = path.join(profilePath, "extensions");
  fs.mkdirSync(extDir, { recursive: true });
  const destXpi = path.join(extDir, `${cfg.capsolver_extension_id}.xpi`);
  if (!fs.existsSync(destXpi)) {
    fs.copyFileSync(CAPSOLVER_XPI_PATH, destXpi);
  }
}

/**
 * Configure CapSolver API key in the extension after browser launch.
 * Reads the Firefox-assigned UUID from extensions-uuid.json, navigates to
 * the extension's options page, and injects the key via chrome.storage / localStorage.
 */
async function configureCapsolverKey(context: BrowserContext, profilePath: string): Promise<void> {
  const cfg = readConfig();
  if (!cfg.capsolver_api_key || !cfg.capsolver_extension_id) return;

  const uuidMapPath = path.join(profilePath, "extensions-uuid.json");
  if (!fs.existsSync(uuidMapPath)) return; // Extension not yet registered by Firefox

  let uuid: string;
  try {
    const uuidMap = JSON.parse(fs.readFileSync(uuidMapPath, "utf-8"));
    uuid = uuidMap[cfg.capsolver_extension_id];
    if (!uuid) return;
  } catch {
    return;
  }

  const optionsPage = await context.newPage();
  try {
    await optionsPage.goto(`moz-extension://${uuid}/options.html`, {
      timeout: 5000,
      waitUntil: "domcontentloaded",
    });
    await optionsPage.evaluate((key: string) => {
      // Try chrome.storage.local first (CapSolver uses Chrome extension APIs),
      // fall back to localStorage as a secondary mechanism.
      const store =
        (typeof chrome !== "undefined" && (chrome as any)?.storage?.local) ||
        (typeof browser !== "undefined" && (browser as any)?.storage?.local);
      if (store) store.set({ apiKey: key });
      localStorage.setItem("capsolver_apikey", key);
      localStorage.setItem("apiKey", key);
    }, cfg.capsolver_api_key);
  } catch {
    // Options page not found or unavailable — non-fatal, user can configure manually
  } finally {
    try { await optionsPage.close(); } catch {}
  }
}

export class BrowserManager {
  refs = new RefRegistry();
  private browser: Browser | null = null;
  private context: BrowserContext | null = null;
  private page: Page | null = null;
  private persistent: string | null;
  private proxy: string | null;
  private history: string[] = [];
  private historyIndex = -1;

  constructor(persistent: string | null = null, proxy: string | null = null) {
    this.persistent = persistent;
    this.proxy = proxy;
  }

  async launch(headless: boolean = true): Promise<void> {
    if (this.browser || this.context) return;

    ensureBrowserInstalled();

    if (this.proxy) {
      if (!this.proxy.includes("://")) {
        throw new Error(
          `Invalid proxy URL: ${this.proxy}. Expected format: http://host:port`
        );
      }
      const scheme = this.proxy.split("://")[0].toLowerCase();
      if (scheme !== "http" && scheme !== "https") {
        throw new Error(
          `Unsupported proxy scheme: ${scheme}. Only http:// and https:// proxies are supported.`
        );
      }
    }

    const launchOpts: Record<string, unknown> = { headless };
    if (this.proxy) launchOpts.proxy = this.proxy;

    if (this.persistent) {
      installCapsolverToProfile(this.persistent);

      const opts = await launchOptions(launchOpts);
      this.context = await firefox.launchPersistentContext(this.persistent, opts);
      const pages = this.context.pages();
      this.page = pages[0] || await this.context.newPage();

      await configureCapsolverKey(this.context, this.persistent);
    } else {
      this.browser = await Camoufox(launchOpts) as Browser;
      this.page = await this.browser.newPage();
      this.context = this.page.context();
    }
  }

  getPage(): Page {
    if (!this.page) throw new Error("Browser not launched. Send 'open' command first.");
    return this.page;
  }

  getContext(): BrowserContext {
    if (!this.context) throw new Error("Browser not launched. Send 'open' command first.");
    return this.context;
  }

  async getTabsAsync(): Promise<{ index: number; url: string; title: string; active: boolean }[]> {
    const ctx = this.getContext();
    const pages = ctx.pages();
    const tabs = [];
    for (let i = 0; i < pages.length; i++) {
      tabs.push({
        index: i,
        url: pages[i].url(),
        title: await pages[i].title(),
        active: pages[i] === this.page,
      });
    }
    return tabs;
  }

  async switchToTab(index: number): Promise<Page> {
    const ctx = this.getContext();
    const pages = ctx.pages();
    if (index < 0 || index >= pages.length) {
      throw new RangeError(`Tab index ${index} out of range (0-${pages.length - 1})`);
    }
    this.page = pages[index];
    await this.page.bringToFront();
    return this.page;
  }

  async closeCurrentTab(): Promise<void> {
    const ctx = this.getContext();
    const pages = ctx.pages();
    if (pages.length <= 1) {
      throw new Error("Cannot close the last tab. Use 'close' to shut down the browser.");
    }
    const current = this.page!;
    const idx = pages.indexOf(current);
    const newIdx = idx > 0 ? idx - 1 : 1;
    this.page = pages[newIdx];
    await this.page.bringToFront();
    await current.close();
  }

  pushHistory(url: string): void {
    this.history = this.history.slice(0, this.historyIndex + 1);
    this.history.push(url);
    this.historyIndex = this.history.length - 1;
  }

  async goBack(): Promise<string | null> {
    if (this.historyIndex <= 0) return null;
    this.historyIndex--;
    const url = this.history[this.historyIndex];
    await this.getPage().goto(url, { waitUntil: "domcontentloaded" });
    return url;
  }

  async goForward(): Promise<string | null> {
    if (this.historyIndex >= this.history.length - 1) return null;
    this.historyIndex++;
    const url = this.history[this.historyIndex];
    await this.getPage().goto(url, { waitUntil: "domcontentloaded" });
    return url;
  }

  async close(): Promise<void> {
    if (this.browser) {
      try { await this.browser.close(); } catch {}
      this.browser = null;
    }
    if (this.context && !this.browser) {
      // persistent context: close context directly
      try { await this.context.close(); } catch {}
    }
    this.context = null;
    this.page = null;
    this.history = [];
    this.historyIndex = -1;
  }

  get isRunning(): boolean {
    return this.browser !== null || this.context !== null;
  }
}
